contextual_word_embeddings.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import os
+from settings import DATA_DIR
+
+
+import numpy as np
+import pandas as pd
+import torch
+from sklearn.preprocessing import StandardScaler
+from transformers import AutoModel, AutoTokenizer
+
+
+def get_embedding(text, tokenizer, model):
+    """
+    Generate the embedding for a given text using the specified tokenizer and model.
+
+    Args:
+        text (str): The input text.
+        tokenizer: The tokenizer instance.
+        model: The model instance.
+
+    Returns:
+        np.ndarray: The embedding vector as a NumPy array.
+    """
+    # Tokenize the input text
+    inputs = tokenizer(
+        text,
+        return_tensors='pt',
+        truncation=True,
+        padding=True,
+        max_length=512  # Set max_length to handle very long texts
+    )
+    # Get the model outputs without tracking gradients
+    with torch.no_grad():
+        outputs = model(**inputs)
+    # Extract the embeddings from the [CLS] token
+    embedding = outputs.last_hidden_state[:, 0, :]
+    return embedding.squeeze().numpy()
+
+
+def combine_embeddings(embeddings, method='concatenate', weights=None):
+    """
+    Combine a list of embeddings into a single embedding vector.
+
+    Args:
+        embeddings (list of np.ndarray): List of embedding vectors.
+        method (str): Method to combine embeddings ('concatenate', 'average', 'weighted').
+        weights (list of float, optional): Weights for weighted sum. Only used if method is 'weighted'.
+
+    Returns:
+        np.ndarray: The combined embedding vector.
+    """
+    if method == 'concatenate':
+        combined_embedding = np.concatenate(embeddings)
+    elif method == 'average':
+        combined_embedding = np.mean(embeddings, axis=0)
+    elif method == 'weighted':
+        if weights is None or len(weights) != len(embeddings):
+            raise ValueError("Weights must be provided and match the number of embeddings for weighted combination.")
+        combined_embedding = np.average(embeddings, axis=0, weights=weights)
+    else:
+        raise ValueError(f"Invalid combination method: {method}")
+    return combined_embedding
+
+
+def main(input_file, output_file, columns, model_name='ufal/robeczech-base', combine_method='concatenate', weights=None, ):
+    """
+    Main function to read input CSV, generate embeddings, and save to output CSV.
+
+    Args:
+        input_file (str): Path to the input CSV file.
+        output_file (str): Path to the output CSV file.
+        model_name (str): Name of the pre-trained model.
```

Comment:
I would prefer to have the args in docstring in the same order as in the function declaration

Reply:
good point